scripts/lib/CIME/XML/generic_xml.py

```diff
@@ -35,31 +57,169 @@ def __init__(self, infile=None, schema=None):
+            else:
+                self.root = self.make_child("file", root=root, attributes={"id":os.path.basename(infile), "version":"2.0"})
+
             self.tree = ET.ElementTree(root)
 
     def read(self, infile, schema=None):
         """
         Read and parse an xml file into the object
         """
-        logger.debug("read: " + infile)
+        logger.debug("read: " + str(infile))
+
         with open(infile, 'r') as fd:
-            if self.tree:
-                self.root.append(ET.parse(fd).getroot())
-            else:
-                self.tree = ET.parse(fd)
-                self.root = self.tree.getroot()
+            self.read_fd(fd)
 
         if schema is not None and self.get_version() > 1.0:
             self.validate_xml_file(infile, schema)
 
         logger.debug("File version is {}".format(str(self.get_version())))
 
+    def read_fd(self, fd):
+        if self.tree:
+            self.add_child(_Element(ET.parse(fd).getroot()))
+        else:
+            self.tree = ET.parse(fd)
+            self.root = _Element(self.tree.getroot())
+
+    #
+    # API for individual node operations
+    #
+
+    def get(self, node, attrib_name, default=None):
+        return node.xml_element.get(attrib_name, default=default)
+
+    def has(self, node, attrib_name):
+        return attrib_name in node.xml_element.attrib
+
+    def set(self, node, attrib_name, value):
+        return node.xml_element.set(attrib_name, value)
+
+    def pop(self, node, attrib_name):
+        return node.xml_element.attrib.pop(attrib_name)
+
+    def attrib(self, node):
+        # Return a COPY. We do not want clients making changes directly
+        return None if node.xml_element.attrib is None else dict(node.xml_element.attrib)
+
+    def set_name(self, node, name):
+        node.xml_element.tag = name
+
+    def set_text(self, node, text):
+        node.xml_element.text = text
+
+    def name(self, node):
+        return node.xml_element.tag
+
+    def text(self, node):
+        return node.xml_element.text
+
+    def add_child(self, node, root=None):
+        """
+        Add element node to self at root
+        """
+        root = root if root is not None else self.root
+        root.xml_element.append(node.xml_element)
+
+    def copy(self, node):
+        return deepcopy(node)
+
+    def remove_child(self, node, root=None):
+        root = root if root is not None else self.root
+        root.xml_element.remove(node.xml_element)
+
+    def make_child(self, name, attributes=None, root=None, text=None):
+        root = root if root is not None else self.root
+        if attributes is None:
+            node = _Element(ET.SubElement(root.xml_element, name))
+        else:
+            node = _Element(ET.SubElement(root.xml_element, name, attrib=attributes))
+
+        if text:
+            self.set_text(node, text)
+
+        return node
+
+    def get_children(self, name=None, attributes=None, root=None, no_validate=False):
```

Comment:
Important function. This will be the core code for searching for direct children of a node.